Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBOAuthTokenRequest.m

```diff
@@ -0,0 +1,180 @@
+
+  request.HTTPMethod = @"POST";
+  request.HTTPBody = paramsData;
+  [request addValue:[DBTransportBaseConfig defaultUserAgent] forHTTPHeaderField:@"User-Agent"];
+  [request addValue:@"application/x-www-form-urlencoded; charset=utf-8" forHTTPHeaderField:@"Content-Type"];
+
+  __weak typeof(self) weakSelf = self;
+  return [[NSURLSession sharedSession] dataTaskWithRequest:request
+                                     completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+    [weakSelf db_handleResponse:response data:data error:error];
+  }];
+
+  return [[NSURLSession sharedSession] dataTaskWithRequest:request];
+}
+
+- (void)db_handleResponse:(NSURLResponse *)response
+                     data:(NSData *)data
+                    error:(NSError *)error {
+  NSDictionary<NSString *, id> *resultDict = [DBOAuthTokenExchangeRequest resultDictionaryFromData:data];
+  DBOAuthResult *result = nil;
+  if (error) {
+    // Network error
+    result = [DBOAuthResult unknownErrorWithErrorDescription:error.localizedDescription];
+  } else {
+    // No network error, parse response data
+    result = [DBOAuthTokenExchangeRequest db_extractResultFromDict:resultDict];
+  }
+
+  dispatch_queue_t queue = _queue ?: dispatch_get_main_queue();
+  dispatch_async(queue, ^{
+    self->_completion(result);
+  });
+}
+
++ (NSDictionary<NSString *, id> *)resultDictionaryFromData:(NSData *)data {
+  id json = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil];
+  if ([json isKindOfClass:[NSDictionary<NSString *, id> class]]) {
+    return json;
+  } else {
+    return nil;
+  }
+}
+
++ (DBOAuthResult *)db_extractResultFromDict:(NSDictionary<NSString *, id> *)dict {
+  // Interpret success result if any.
+  DBOAuthResult *successResult = [self db_extractSuccessResultFromDict:dict];
+  if (successResult) {
+    return successResult;
+  }
+
+  // Interpret error if any.
+  DBOAuthResult *errorResult = [self db_extractErrorResultFromDict:dict];
+  if (errorResult) {
+    return errorResult;
+  }
+
+  return [DBOAuthResult unknownErrorWithErrorDescription:@"Invalid response."];
+}
+
++ (DBOAuthResult *)db_extractErrorResultFromDict:(NSDictionary<NSString *, id> *)dict {
+  id errorCode = dict[@"error"];
+  if (!errorCode) {
+    return nil;
+  }
+  id errorDescription = dict[@"error_description"];
+  if ([errorCode isKindOfClass:[NSString class]] && [errorDescription isKindOfClass:[NSString class]]) {
+    return [[DBOAuthResult alloc] initWithError:errorCode errorDescription:errorDescription];
+  } else {
+    return [DBOAuthResult unknownErrorWithErrorDescription:nil];
+  }
+}
+
++ (DBOAuthResult *)db_extractSuccessResultFromDict:(NSDictionary<NSString *, id> *)dict {
+  id tokenType = dict[@"token_type"];
+  id accessToken = dict[@"access_token"];
+  id refreshToken = dict[@"refresh_token"];
+  id userId = dict[@"uid"];
+  id expiresIn = dict[@"expires_in"];
+
+  BOOL valid = true;
```

Comment:
BOOL is an unsigned char. to be consistent, either use YES or NO, or use the `bool` type. 

Also, this boolean is modified via a bitwise operation which is technically incorrect for booleans. We should use `&&` or `||`